ulib/FStar.List.Pure.Properties.fst

```diff
@@ -52,3 +52,200 @@ let rec splitAt_length_total (#a:Type) (l:list a)
   match l with
   | [] -> ()
   | x :: xs -> splitAt_length_total xs
+
+
+(** If we [append] the two lists produced using a [splitAt], then we
+    get back the original list *)
+let rec lemma_splitAt_append (#a:Type) (n:nat) (l:list a) :
+  Lemma
+    (requires n <= length l)
+    (ensures (let l1, l2 = splitAt n l in
+              append l1 l2 == l)) =
+  match n with
+  | 0 -> ()
+  | _ ->
+    match l with
+    | [] -> ()
+    | x :: xs -> lemma_splitAt_append (n-1) xs
+
+
+(** If we [splitAt] the point at which two lists have been [append]ed, then we
+    get back the original lists. *)
+let rec lemma_append_splitAt (#t:Type) (l1 l2:list t) :
+  Lemma
+    (ensures (splitAt (length l1) (append l1 l2) == (l1, l2))) =
+  match l1 with
+  | [] -> ()
+  | _ -> lemma_append_splitAt (tl l1) l2
+
+
+(** The [hd] of the second list returned via [splitAt] is the [n]th element of
+    the original list *)
+let rec lemma_splitAt_index_hd (#t:Type) (n:nat) (l:list t) :
+  Lemma
+    (requires (n < length l))
+    (ensures (let l1, l2 = splitAt n l in
+              splitAt_length n l;
+              length l2 > 0 /\ hd l2 == index l n)) =
+  let x :: xs = l in
+  match n with
+  | 0 -> ()
+  | _ -> lemma_splitAt_index_hd (n - 1) (tl l)
+
+
+(** If two lists have the same left prefix, then shorter left prefixes are
+    also the same. *)
+let rec lemma_splitAt_shorten_left
+    (#t:Type) (l1 l2:list t) (i:nat{i <= length l1 /\ i <= length l2}) (j:nat{j <= i}) :
+  Lemma
+    (requires (fst (splitAt i l1) == fst (splitAt i l2)))
+    (ensures (fst (splitAt j l1) == fst (splitAt j l2))) =
+  match j with
+  | 0 -> ()
+  | _ ->
+    lemma_splitAt_shorten_left (tl l1) (tl l2) (i-1) (j-1)
+
+(** Doing an [index] on the left-part of a [splitAt] is same as doing it on
+    the original list *)
+let rec lemma_splitAt_reindex_left (#t:Type) (i:nat) (l:list t) (j:nat) :
+  Lemma
+    (requires i <= length l /\ j < i)
+    (ensures (
+        let left, right = splitAt i l in
+        splitAt_length i l;
+        j < length left /\ index left j == index l j)) =
+  match i, j with
+  | 1, _ | _, 0 -> ()
+  | _ -> lemma_splitAt_reindex_left (i - 1) (tl l) (j - 1)
+
+
+(** Doing an [index] on the right-part of a [splitAt] is same as doing it on
+    the original list, but shifted *)
+let rec lemma_splitAt_reindex_right (#t:Type) (i:nat) (l:list t) (j:nat) :
+  Lemma
+    (requires i <= length l /\ j + i < length l)
+    (ensures (
+        let left, right = splitAt i l in
+        splitAt_length i l;
+        j < length right /\ index right j == index l (j + i))) =
+  match i with
+  | 0 -> ()
+  | _ -> lemma_splitAt_reindex_right (i - 1) (tl l) j
+
```

Comment:
`splitAt n l` is unusual in that it's defined for all `n`. Was that intentional @kyoDralliam? Why not define it only for `n <= length l` or maybe even `n < length l`? That's a separate question independent of this PR.

Now, I was expecting to see the library provide a single lemma that characterizes the equational behavior of splitAt in terms of other existing notions for which we already have lemmas. For instance, a lemma like:

```
n <= length l ==>
(splitAt n l = l1, l2 <==>
l == l1 @ l2 /\
length l1 = n)
```

`lemma_splitAt_append` and `lemma_append_splitAt` come close to this, but do not quite constrain `n` sufficiently.

Reply:
I'm not sure I understand what the "`lemma_splitAt_append` and `lemma_append_splitAt` come close to this, but do not quite constrain `n` sufficiently" means. 

Since `n` is a `nat`, it is always `>=0`. Additionally, `lemma_splitAt_append` has a `requires` clause of `n <= length l`. I believe that this constraint (rather than `n < length l`) makes sense. Consider this example:

```
l = [ 0 ; 1 ; 2 ; 3 ]
//  ^   ^   ^   ^   ^
//  0   1   2   3   4
```

We would want all these possible points for splitting (i.e., `0 <= n <= length l`).

Also, should I merge the lemmas as per what you've said, or should I leave that for a separate PR?

Reply:
This is what I mean about "...not quite constrain `n` sufficiently"

```
--- a/ulib/FStar.List.Pure.Properties.fst
+++ b/ulib/FStar.List.Pure.Properties.fst
@@ -60,7 +60,7 @@ let rec lemma_splitAt_append (#a:Type) (n:nat) (l:list a) :
   Lemma
     (requires n <= length l)
     (ensures (let l1, l2 = splitAt n l in
-              append l1 l2 == l)) =
+              append l1 l2 == l /\ n = length l1)) =
   match n with
   | 0 -> ()
   | _ ->
@@ -78,6 +78,10 @@ let rec lemma_append_splitAt (#t:Type) (l1 l2:list t) :
   | [] -> ()
   | _ -> lemma_append_splitAt (tl l1) l2

+let rec splitAt_append (#t:Type) (l l1 l2:list t) (n:nat{n <= length l})
+  : Lemma (splitAt n l == (l1, l2) <==> l == l1 @ l2 /\ length l1 = n)
+  = lemma_splitAt_append n l;
+    lemma_append_splitAt l1 l2
```

I think a single `splitAt_append` lemma that fully characterizes the behavior of `splitAt n l` in terms of more standard list concepts like `@` and `length` would be useful. I've shown this at least for ` n <= length l`, but you could easily also prove a more general lemma (where if `n > length l` then `l2 = []`)

Reply:
This is already covered by the pre-existing `splitAt_length`, which is why I didn't include it here.

Reply:
Added this in.